src/jaybenne/sourcing.cpp

```diff
@@ -87,12 +101,37 @@ TaskStatus SourcePhotons(T *md, const Real t_start, const Real dt) {
               [[maybe_unused]] const Real &sbd = sb;
               [[maybe_unused]] const Real &vvd = vv;
               [[maybe_unused]] const Real &dtd = dt;
-              [[maybe_unused]] auto &opac = opacity;
+              [[maybe_unused]] auto *opac = opacity;
+              [[maybe_unused]] auto *mopac = mopacity;
+              [[maybe_unused]] const auto numind = numin;
+              [[maybe_unused]] const auto numaxd = numax;
+              [[maybe_unused]] const auto n_nubinsd = n_nubins;
               Real erad = 0.0;
               if constexpr (ST == SourceType::thermal) {
                 erad = (4.0 * sbd / vvd) * std::pow(temp, 4.0) * dv;
               } else if constexpr (ST == SourceType::emission) {
-                const Real emis = opac.Emissivity(rho, temp);
+                Real emis = -1.;
+                if constexpr (FT == FrequencyType::gray) {
+                  emis = opac->Emissivity(rho, temp);
+                } else if constexpr (FT == FrequencyType::multigroup) {
+                  // Construct emission CDF
+                  const Real dlnu = (std::log(numaxd) - std::log(numind)) / n_nubinsd;
+                  vmesh(b, fj::emission_cdf(0), k, j, i) = opac->EmissivityPerNu(
+                      rho, temp, std::exp(std::log(numind) + 0.5 * dlnu));
+                  for (int n = 1; n < n_nubinsd; n++) {
+                    const Real nu = std::exp(std::log(numind) + (n + 0.5) * dlnu);
+                    vmesh(b, fj::emission_cdf(n), k, j, i) =
+                        opac->EmissivityPerNu(rho, temp, nu) +
+                        vmesh(b, fj::emission_cdf(n - 1), k, j, i);
+                  }
+                  for (int n = 0; n < n_nubinsd; n++) {
+                    // Get total emissivity
+                    emis += vmesh(b, fj::emission_cdf(n), k, j, i);
+                    // Normalize emission CDF
+                    vmesh(b, fj::emission_cdf(n), k, j, i) /=
+                        vmesh(b, fj::emission_cdf(n_nubins - 1), k, j, i);
```

Comment:
```suggestion
                        vmesh(b, fj::emission_cdf(n_nubinsd - 1), k, j, i);
```
? Also, should the normalization occur after the aggregation?

Reply:
Yes good catch with the `n_nubinsd`. 

I think the normalization is properly done after aggregation because the previous loop is like `cdf(n) = emis + cdf(n-1)` right? Maybe I'm missing something

Reply:
I couldn't see where `emis` was getting used to set the `fj::emission_cdf` values in this or the previous loops... I could certainly be missing something.

Reply:
Wait - nevermind - I see where `emission_cdf` is getting set above.